types/sanctuary/index.d.ts

```diff
@@ -186,98 +179,106 @@ declare namespace Sanctuary {
     pipe<A, B, C, D>(fs: [Fn<A, B>, Fn<B, C>, Fn<C, D>]): (x: A) => D;
     pipe<A, B, C, D, E>(fs: [Fn<A, B>, Fn<B, C>, Fn<C, D>, Fn<D, E>]): (x: A) => E;
     pipe<A, B, C, D, E, F>(fs: [Fn<A, B>, Fn<B, C>, Fn<C, D>, Fn<D, E>, Fn<E, F>]): (x: A) => F;
-    pipe(fs: ReadonlyArray<Fn<any, any>>): (x: any) => any;
+    pipe(tk: TK): (tk: TK) => TK;
+    pipeK(tk: TK): (tk: TK) => TK;
     on<A, B, C>(p: Fn2<B, B, C>): (q: Fn<A, B>) => (r: A) => Fn<A, C>;
-    //  TODO: Maybe
+    //  Pair
+    Pair<A>(x: A): <B>(y: B) => Pair<A, B>;
+    pair<A, B, C>(f: Fn2<A, B, C>): (p: Pair<A, B>) => C;
```

Comment:
@vicrac, is it possible to get type inference to work here so one need not do the following?

```typescript
S.pair ((x: string) => (y: string) => x + y) (S.Pair ('foo') ('bar'))
```

My understanding is that TypeScript's left-to-right resolution algorithm gives `A` = :woman_shrugging:, `B` = :woman_shrugging:, `C` = :woman_shrugging:. I don't think there's anything we can do about this, but I would love to be wrong.

Reply:
Unfortunately, no idea. Perhaps it is possible to at least extract these types to make them explicit type parameters (`S.pair<string, string>(x => y => x+y)`) but then we'd need to provide third parameter too, and I don't know how to make Typescript infer it for us (althought there sure is some way)

Reply:
As a fairly new TypeScript user I'm getting used to giving `tsc` hints. Hopefully users won't be frustrated by the number of hints required to use Sanctuary with TypeScript.

Reply:
Perhaps there's some better way, but I am getting quite confused on whether such a type inference is _actually even possible_ with curried functions.

Reply:
Possible in general, or possible in TypeScript? It's certainly possible in general, but having read microsoft/TypeScript#15680 and similar issues it seems highly unlikely that TypeScript will ever be capable of correctly inferring the type of an expression involving polymorphic higher-order functions. TypeScript's type inference is intentionally limited.